### PR TITLE
Rev Components are no longer leaked + Rev and Zombie icon visibility to ghosts is now controlled by a cvar

### PR DESCRIPTION
**Pull request by nikthechampiongr**
## About the PR

Partially addresses #21529.

This PR does 3 main things:
1. Makes it so the Revolutionary and HeadRevolutionary components do not get shared with clients that don't need them.
2. Creates the ghost.rev_icons_visible_to_ghosts and ghost.zombie_icons_visible_to_ghosts Cvars which determine whether ghosts can see status icons on Revs and Zombies respectively.
3. Creates the ShowRevIcons Component and the ShowZombieIcons Component which allow someone to always view the status icons for revs and zombies respectively.

It should be noted that this does not touch the similar but different system introduced in #21260. That is out of scope for this pr.

## Why / Balance

As I understand it, there is concern that hacked clients can currently "know" someone's antag status because the relevant components are shared with all clients. This partially addresses this issue by making it so the components for revs are not shared unless a client needs to know about them. This is not a complete fix and I am sure there are probably other components that may rat out someone's antag status beyond the rev components but it is a start.

This additionally gives servers more power to determine who can see the status of team antags. Currently it's hard coded that ghosts are able to see revs and zombies. Now if a server does not deem it desirable, they can disable this behavior. The default behavior has not changed.
## Technical details

The biggest change is that now the Rev component and Head rev component are SessionSpecific. The SharedRevolutionarySystem now determines whether to send them to clients or not. 

For cases where a client later gains the ability to view the icons (For example the player is converted) I am forced to dirty all rev/head rev components in order to retry sending them. This is not very efficient but to my knowledge there is no way to dirty them specifically for one client. 

It should be noted that this still allows for clients that should no longer be able to know about the components(for example if the player gets deconverted) but since that is controlled by the client, and either way the player themselves have a brain to remember things with I saw no point in trying to make sure they are immediately deleted on the client's side.

The ZombieComponent was not made SessionSpecific since it's needed by clients to change the appearance of zombies and either way people should be able to just see when someone is a zombie. If you use a hacked client to display the icons anyways that's just a skill issue.

Other than that, the Zombie and Revolutionary system on client no longer inherit from the AntagStatusIconSystem, and are not responsible for receiving the GetStatusIconsEvent as this responsibility has been transferred to the AntagStatusIconSystem. They do however now listen for CanDisplayStatusIconsEvent which they can cancel so that the icon is not displayed.

The Interface IAntagStatusIconComponent was added so I could use generics effectively in the AntagStatusIconSystem. It just gives a common interface for getting the StatusIcon from a component.
## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

https://github.com/space-wizards/space-station-14/assets/32041239/58a65344-0a80-4c90-a073-8bd458b93045

https://github.com/space-wizards/space-station-14/assets/32041239/6ea975a1-1646-4c4b-88d1-666bffca92ed

## Breaking changes

The AntagStatusIconsSystem now directly handles GetStatusIcons events. Most of the logic on whether to add an icon should now be handled through the CanDisplayStatusIconsEvent.

The AntagStatusIConsSystem is now sealed, not abstract, and the revolutionary and zombie systems on client no longer inherit from it.

If you were relying on clients always having knowledge of rev components for some reason then this is also a breaking change.

**Changelog**

no cl no fun

### CODE REVIEW

**Comment by notquitehadouken** [comment]
pvs success

**Comment by nikthechampiongr** [comment]
> pvs success

Truly

File: Content.Client/Revolutionary/RevolutionarySystem.cs [hunk 1]
```diff
@@ -1,35 +1,60 @@
...
+
+    private bool _revIconGhostVisibility;
     public override void Initialize()
     {
         base.Initialize();
 
-        SubscribeLocalEvent<RevolutionaryComponent, GetStatusIconsEvent>(GetRevIcon);
-        SubscribeLocalEvent<HeadRevolutionaryComponent, GetStatusIconsEvent>(GetHeadRevIcon);
+        _cfg.OnValueChanged(CCVars.RevIconsVisibleToGhosts, value => _revIconGhostVisibility = value, true);
+        SubscribeLocalEvent<RevolutionaryComponent, CanDisplayStatusIconsEvent>(OnCanShowRevIcon);
```
Comment (metalgearsloth) [inline Content.Client/Revolutionary/RevolutionarySystem.cs:24]: This event will leak, you need to unsubscribe on the client.

File: Content.Shared/Revolutionary/SharedRevolutionarySystem.cs [hunk 2]
```diff
@@ -1,20 +1,34 @@
...
     [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
     [Dependency] private readonly SharedStunSystem _sharedStun = default!;
+    [Dependency] private readonly IConfigurationManager _cfg = default!;
 
+    private bool _revIconGhostVisibility;
     public override void Initialize()
     {
         base.Initialize();
+        _cfg.OnValueChanged(CCVars.RevIconsVisibleToGhosts, value => _revIconGhostVisibility = value, true);
+
```
Comment (metalgearsloth) [inline Content.Shared/Revolutionary/SharedRevolutionarySystem.cs:25]: Same here.

File: Content.Client/Zombies/ZombieSystem.cs [hunk 3]
```diff
@@ -1,21 +1,28 @@
...
+    [Dependency] private readonly IConfigurationManager _cfg = default!;
+
+    private bool _zombieIconGhostVisibility;
 
     public override void Initialize()
     {
         base.Initialize();
 
+        _cfg.OnValueChanged(CCVars.ZombieIconsVisibleToGhosts, value => _zombieIconGhostVisibility= value, true);
         SubscribeLocalEvent<ZombieComponent, ComponentStartup>(OnStartup);
```
Comment (metalgearsloth) [inline Content.Client/Zombies/ZombieSystem.cs:24]: Here as well.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
My worry is the duplication between revs and zombies may lead to further copying down the line.

**Comment by nikthechampiongr** [comment]
> My worry is the duplication between revs and zombies may lead to further copying down the line.

I cut down on the duplication on the AntagStatusIconSystem. For the leaking stuff, I asked for a clarification on what you mean on discord.

File: Content.Client/Revolutionary/RevolutionarySystem.cs (hunk 1 above)
Comment (nikthechampiongr) [reply Content.Client/Revolutionary/RevolutionarySystem.cs:24]: This is the client. This event is needed to actually draw the icons.

**Comment by nikthechampiongr** [comment]
how the hell is robust toolbox making it in there.

File: Content.Shared/Revolutionary/SharedRevolutionarySystem.cs [hunk 4]
```diff
@@ -37,4 +46,64 @@ private void MindShieldImplanted(EntityUid uid, MindShieldComponent comp, MapIni
...
+        {
+            Dirty(uid, comp);
+        }
+
+        var headRevComps = EntityQueryEnumerator<HeadRevolutionaryComponent>();
+        while (headRevComps.MoveNext(out var uid, out var comp))
+        {
+            Dirty(uid, comp);
+        }
+    }
```
Comment (metalgearsloth) [inline Content.Shared/Revolutionary/SharedRevolutionarySystem.cs:108]: AllEntityQuery instead to include paused ents (which you should).

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**